src/comgr.cpp

```diff
@@ -0,0 +1,503 @@
+        CASE_TO_STRING(AMD_COMGR_DATA_KIND_DIAGNOSTIC);
+        CASE_TO_STRING(AMD_COMGR_DATA_KIND_LOG);
+        CASE_TO_STRING(AMD_COMGR_DATA_KIND_BC);
+        CASE_TO_STRING(AMD_COMGR_DATA_KIND_RELOCATABLE);
+        CASE_TO_STRING(AMD_COMGR_DATA_KIND_EXECUTABLE);
+        CASE_TO_STRING(AMD_COMGR_DATA_KIND_BYTES);
+        CASE_TO_STRING(AMD_COMGR_DATA_KIND_FATBIN);
+    }
+    return "<Unknown data kind>";
+}
+
+static std::string to_string(const amd_comgr_action_kind_t val)
+{
+    switch(val)
+    {
+        CASE_TO_STRING(AMD_COMGR_ACTION_SOURCE_TO_PREPROCESSOR);
+        CASE_TO_STRING(AMD_COMGR_ACTION_ADD_PRECOMPILED_HEADERS);
+        CASE_TO_STRING(AMD_COMGR_ACTION_COMPILE_SOURCE_TO_BC);
+        CASE_TO_STRING(AMD_COMGR_ACTION_ADD_DEVICE_LIBRARIES);
+        CASE_TO_STRING(AMD_COMGR_ACTION_LINK_BC_TO_BC);
+        CASE_TO_STRING(AMD_COMGR_ACTION_OPTIMIZE_BC_TO_BC);
+        CASE_TO_STRING(AMD_COMGR_ACTION_CODEGEN_BC_TO_RELOCATABLE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_CODEGEN_BC_TO_ASSEMBLY);
+        CASE_TO_STRING(AMD_COMGR_ACTION_LINK_RELOCATABLE_TO_RELOCATABLE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_LINK_RELOCATABLE_TO_EXECUTABLE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_ASSEMBLE_SOURCE_TO_RELOCATABLE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_DISASSEMBLE_RELOCATABLE_TO_SOURCE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_DISASSEMBLE_EXECUTABLE_TO_SOURCE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_DISASSEMBLE_BYTES_TO_SOURCE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_COMPILE_SOURCE_TO_FATBIN);
+    }
+    return "<Unknown action kind>";
+}
+
+static bool PrintVersion()
+{
+    std::size_t major = 0;
+    std::size_t minor = 0;
+    (void)amd_comgr_get_version(&major, &minor);
+    MIOPEN_LOG_I("comgr v." << major << '.' << minor);
+    return true;
+}
+
+static std::string GetStatusText(const amd_comgr_status_t status)
+{
+    const char* reason = nullptr;
+    if(AMD_COMGR_STATUS_SUCCESS != amd_comgr_status_string(status, &reason))
+        reason = "<Unknown>";
+    return std::string(reason) + " (" + std::to_string(static_cast<int>(status)) + ')';
+}
+
+static void LogOptions(const char* options[], size_t count)
+{
+#if DEBUG_DETAILED_LOG
+    if(miopen::IsLogging(miopen::LoggingLevel::Info))
+    {
+        std::ostringstream oss;
+        for(std::size_t i = 0; i < count; ++i)
+            oss << options[i] << '\t';
+        MIOPEN_LOG_I(oss.str());
+    }
+#else
+    std::ignore = options;
+    std::ignore = count;
+#endif
+}
+
+class Dataset;
+static std::string GetLog(const Dataset& dataset, bool catch_comgr_exceptions = false);
+
+struct ComgrError : std::exception
+{
+    amd_comgr_status_t status;
+    std::string text;
+
+    ComgrError(const amd_comgr_status_t s) : status(s) {}
+    ComgrError(const amd_comgr_status_t s, const std::string& t) : status(s), text(t) {}
+    const char* what() const noexcept override { return text.c_str(); }
+};
+
+struct ComgrOwner
```

Comment:
This base class seems pointless. The base methods can be made static/free functions and the constructors are not needed if we use `MIOPEN_MANAGE_PTR`.

Reply:
[duplicate]